packages/core/src/scuttle.js

```diff
@@ -0,0 +1,181 @@
+  getOwnPropertyNames,
+  getOwnPropertyDescriptor,
+  create,
+  defineProperty,
+} = Object
+
+const { isArray, from } = Array
+
+const { getPrototypeOf } = Reflect
+
+const { warn } = console
+
+/**
+ * Applies scuttling, with the default set of options, including using Snow if passed in as scuttlerFunc.
+ * Scuttle globalThis right after we used it to create the root package compartment.
+ *
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {ScuttleOpts} opts - Scuttling options.
+ */
+function scuttle(globalRef, opts) {
+  const scuttleOpts = generateScuttleOpts(globalRef, opts)
+
+  if (scuttleOpts.enabled) {
+    if (!isArray(scuttleOpts.exceptions)) {
+      throw new Error(
+        `LavaMoat - scuttleGlobalThis.exceptions must be an array, got "${typeof scuttleOpts.exceptions}"`
+      )
+    }
+    scuttleOpts.scuttlerFunc(globalRef, (realm) =>
+      performScuttleGlobalThis(realm, scuttleOpts.exceptions)
+    )
+  }
+}
+
+/**
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {ScuttleOpts|boolean} originalOpts - Scuttling options. Accepts `true` for backwards compatibility.
+ * @returns {ScuttleOpts} - Final scuttling options.
+ */
+function generateScuttleOpts(globalRef, originalOpts = create(null)) {
+  const defaultOpts = {
+    enabled: true,
+    exceptions: [],
+    scuttlerName: '',
+  }
+  const opts = assign(
+    {},
+    originalOpts === true ? { ...defaultOpts } : { ...originalOpts },
+    { scuttlerFunc: (globalRef, scuttle) => scuttle(globalRef) },
+    {
+      exceptions: (originalOpts.exceptions || defaultOpts.exceptions).map((e) =>
+        toRE(e)
+      ),
+    }
+  )
+  if (opts.scuttlerName) {
+    if (!globalRef[opts.scuttlerName]) {
+      throw new Error(
+        `LavaMoat - 'scuttlerName' function "${opts.scuttlerName}" expected on globalRef.` +
+          'To learn more visit https://github.com/LavaMoat/LavaMoat/pull/462.'
+      )
+    }
+    opts.scuttlerFunc = globalRef[opts.scuttlerName]
+  }
+  return opts
+
+  /**
+   * @param {string|RegExp} except - Exception to convert to RegExp.
+   * @returns {string|RegExp} - Converted exception.
+   */
+  function toRE(except) {
+    // turn scuttleGlobalThis.exceptions regexes strings to actual regexes
+    if (!except.startsWith('/')) {
+      return except
+    }
+    const parts = except.split('/')
+    const pattern = parts.slice(1, -1).join('/')
+    const flags = parts[parts.length - 1]
+    return new RegExp(pattern, flags)
+  }
+}
+
+/**
+ * Runs scuttling on the globalRef. Use applyDefaultScuttling for full scope of options.
+ *
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {Array<string|RegExp>} extraPropsToAvoid - List of additional properties to exclude from scuttling beyond the default ones.
+ */
+function performScuttleGlobalThis(globalRef, extraPropsToAvoid = new Array()) {
+  const props = new Array()
+  getPrototypeChain(globalRef).forEach((proto) =>
```

Comment:
our iteration is inconsistent.  why not a `.map()`?  why use a `for...of` loop below?

Reply:
fair enough a3d5942e7